kernel/aster-nix/src/syscall/utimens.rs

```diff
@@ -2,90 +2,229 @@
+use super::{constants::MAX_FILENAME_LEN, SyscallReturn};
 use crate::{
-    fs::{file_table::FileDesc, fs_resolver::FsPath},
+    fs::{
+        file_table::FileDesc,
+        fs_resolver::{FsPath, AT_FDCWD},
+        path::Dentry,
+    },
     prelude::*,
-    syscall::constants::MAX_FILENAME_LEN,
-    time::timespec_t,
+    time::{clocks::RealTimeCoarseClock, timespec_t, timeval_t},
     util::{read_cstring_from_user, read_val_from_user},
 };
 
+/// The 'sys_utimensat' system call sets the access and modification times of a file specified by `pathname`.
+/// The file is specified by a file descriptor `dirfd` instead of a path name.
+/// The times are specified by a timespec structure. times[0] is the access time, and times[1] is the modification time.
+/// The `flags` argument is a bit mask that can include the following values:
+/// - `AT_SYMLINK_NOFOLLOW`: If set, the file is not dereferenced if it is a symbolic link.
 pub fn sys_utimensat(
     dirfd: FileDesc,
-    path_addr: Vaddr,
+    pathname_ptr: Vaddr,
     timespecs_ptr: Vaddr,
     flags: u32,
 ) -> Result<SyscallReturn> {
-    let path = read_cstring_from_user(path_addr, MAX_FILENAME_LEN)?;
-    let (atime, mtime) = {
-        let (autime, mutime) = if timespecs_ptr == 0 {
-            (timespec_t::utime_now(), timespec_t::utime_now())
-        } else {
-            let mut timespecs_addr = timespecs_ptr;
-            let autime = read_val_from_user::<timespec_t>(timespecs_addr)?;
-            timespecs_addr += core::mem::size_of::<timespec_t>();
-            let mutime = read_val_from_user::<timespec_t>(timespecs_addr)?;
-            (autime, mutime)
-        };
-
-        // TODO: Get current time
-        let current_time: timespec_t = Default::default();
-
-        let atime = if autime.is_utime_omit() {
-            None
-        } else if autime.is_utime_now() {
-            Some(current_time)
-        } else {
-            Some(autime)
-        };
-        let mtime = if mutime.is_utime_omit() {
-            None
-        } else if mutime.is_utime_now() {
-            Some(current_time)
-        } else {
-            Some(mutime)
-        };
-        (atime, mtime)
+    debug!(
+        "utimensat: dirfd: {}, pathname_ptr: {:#x}, timespecs_ptr: {:#x}, flags: {:#x}",
+        dirfd, pathname_ptr, timespecs_ptr, flags
+    );
+    let times = if timespecs_ptr != 0 {
+        let (autime, mutime) = read_time_from_user::<timespec_t>(timespecs_ptr)?;
+        if autime.is_utime_omit() && mutime.is_utime_omit() {
+            return Ok(SyscallReturn::Return(0));
+        }
+        Some(Times {
+            atime: autime,
+            mtime: mutime,
+        })
+    } else {
+        None
     };
-    let flags = UtimensFlags::from_bits(flags)
-        .ok_or(Error::with_message(Errno::EINVAL, "invalid flags"))?;
+    do_utimes(dirfd, pathname_ptr, times, flags)
+}
+
+/// The 'sys_futimesat' system call sets the access and modification times of a file specified by `pathname`.
```

Comment:
The 'sys_futimesat' system call sets the access and modification times of a file.